test/e2e/framework/device_plugin_utils.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+	"k8s.io/kubernetes/test/e2e_node/builder"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	// dummy device file saved in "/tmp/dummy-device" directory,
+	// e.g. "/tmp/dummy-device/device-1", "/tmp/dummy-device/device-2"
+	DummyDeviceDir string = "dummy-device"
+	DummyDeviceRE  string = "^device-[0-9]*$"
+)
+
+var (
+	// Dummy device resource name to be advertised to Kubelet
+	ResourceName = "resource.name/1"
+)
+
+func NumberOfStubDevices(node *v1.Node) int64 {
+	val, ok := node.Status.Capacity[v1.ResourceName(ResourceName)]
+
+	if !ok {
+		return 0
+	}
+
+	return val.Value()
+}
+
+// StubDevicePlugin returns the stub Device Plugin pod for conformance testing
+func StubDevicePlugin(ns string) *v1.Pod {
```

Comment:
Can we also pass in config file as a parameter to this function and then pass it as an Arg to the stub-device-plugin container?

Reply:
HI, I think there is no need for `stub-device-plugin` container to know what is `config file`.
It is just a slave to read and watch `dummy device files` of a specific resource and report to Kubelet.

Reply:
Any reason we want to have this slave model instead of just having stub device plugin to execute according to its config file? If we have test to create dummy device files, how are we going to use it in e2e test that can have multiple nodes running device plugin but the test would actually run outside the test cluster?

Reply:
@jiayingz 
Hi, I think there may be two reasons about this:
- one of the main reasons is how to pass config file into stub-device-plugin container without rebuilding image.
- another reason is how to change devices' state (e.g. Healthy -> Unhealthy) without restarting stub-device-plugin if test don't know the config file.

In this slave model, test can write "Unhealthy" into dummy device file to tell stub-device-plugin one device failed.

Considering about e2e test in cluster, I think we can pass it as a `JSON string` instead of a config file. And this `JSON string` is a part of the whole config file which represents a specific resource.

There will be more than one resource in the whole config file, but one stub-device-plugin can only represents one resource for now. The test will parse the whole config file to get how many resources we want to simulate in this test, then determine to create some stub-device-plugins as much as the resources.

But i have no idea about how to simulate device failure in this case :)